aten/src/ATen/InferSize.h

```diff
@@ -37,31 +39,39 @@ inline void infer_size_impl(
     }
   }
 
-  if (TORCH_GUARD_SIZE_OBLIVIOUS(sym_eq(numel, newsize)) ||
-      (infer_dim && newsize > 0 && numel % newsize == 0)) {
-    if (infer_dim) {
-      // We have a degree of freedom here to select the dimension size; follow
-      // NumPy semantics and just bail.  However, a nice error message is needed
-      // because users often use `view` as a way to flatten & unflatten
-      // dimensions and will otherwise be confused why
-      //   empty_tensor.view( 0, 0)
-      // works yet
-      //   empty_tensor.view(-1, 0)
-      // doesn't.
-      TORCH_CHECK(
-          newsize != 0,
-          "cannot reshape tensor of 0 elements into shape ",
-          shape,
-          " because the unspecified dimension size -1 can be any "
-          "value and is ambiguous");
-      res[*infer_dim] = numel / newsize;
-    }
+  auto set_infer_dim = [&]() {
+    // We have a degree of freedom here to select the dimension size; follow
+    // NumPy semantics and just bail.  However, a nice error message is needed
```

Comment:
I found this a bit hard to read. How about

```
    // In the face of dimension size ambiguity, we opt to follow numpy semantics and insert a runtime assert that errors at runtime if we see a tensor with no elements.
```

The whole nice error message part seems self evident?

Reply:
this is exisitng comment i did not add it. its moved as is

Reply:
I see, it's fine to land as is I guess. Bonus points if you clean up the comment to be more understandable.

Reply:
I can do that.